src/client/dfuse/dfuse.h

```diff
@@ -138,6 +138,75 @@ struct dfuse_obj_hdl {
 	bool                      doh_kreaddir_finished;
 };
 
+/* Readdir support.
+ *
+ * Readdir is by far the most complicated component of dfuse as a result of the kernel interfaces,
+ * the dfs interface and the lack of kernel caching for concurrent operations.
+ *
+ * The kernel interface to readdir is to make a callback into dfuse to request a number of dentries
+ * which are then populated in a buffer which is returned to the kernel, each entry in the buffer
+ * contains the name of the entry, the type and some other metadata.  The kernel requests a buffer
+ * size and the dfuse can reply with less if it choses - 0 is taken as end-of-directory.  The
+ * length of the filenames affects the number of entries that will fit in the buffer.
+ *
+ * The dfs interface to reading entries is to call dfs_iterate() which then calls a dfuse callback
+ * with the name of each entry, dfuse then has to perform a lookup to get any metadata for the
+ * entry which it does after the iterate has completed.  DFS takes in a buffer size and max count
+ * which can be much larger than the 4k buffer the kernel uses.
```

Comment:
> dfuse then has to perform a lookup to get any metadata for the entry which it does after the iterate has completed

Nit - Recommend wording to correspond to linear progression. E.g.
After the iterate completes, dfuse performs a lookup to get any ...

> can be much larger than the 4k buffer the kernel uses

So what does this mean for readdir caching? Do we only pass in a 4K buffer to DFS or do we pass in a larger buffer and cache the results?

Reply:
The buffers are packed differently so 4K != 4K in terms of capacity.  dfs_iterate takes a count and a size so dfuse passes in count * (max_size + 1) so the buffer will be big enough for count no matter the file length.

I've updated this comment to say that dfuse will fetch up to 1024 entries from dfs at a time.